core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java

```diff
@@ -210,6 +224,113 @@ private void startCodeGenWatcher(QuarkusClassLoader classLoader, List<CodeGenDat
+        var arthasEnabledOpt = runner.getConfigValue("quarkus.arthas.enabled", Boolean.class);
+        if (arthasEnabledOpt.isEmpty() || !arthasEnabledOpt.get()) {
+            return;
+        }
+
+        var parentDirOpt = runner.getConfigValue("quarkus.arthas.jar-directory", String.class);
+        if (parentDirOpt.isEmpty()) {
+            // should not happen
+            return;
+        }
+        var parentDir = Paths.get(parentDirOpt.get());
+        if (!Files.exists(parentDir)) {
+            Files.createDirectory(parentDir);
+        }
+
+        var arthasHttpPortOpt = runner.getConfigValue("quarkus.arthas.http-port", Integer.class);
+        if (arthasHttpPortOpt.isEmpty()) {
+            // should not happen
+            return;
+        }
+        var arthasHttpPort = arthasHttpPortOpt.get();
+
+        var arthasVersionOpt = runner.getConfigValue("quarkus.arthas.version", String.class);
+        if (arthasVersionOpt.isEmpty()) {
+            // should not happen
+            return;
+        }
+        var arthasVersion = arthasVersionOpt.get();
+        var arthasDir = parentDir.resolve(String.format("arthas-%s", arthasVersion));
+        if (!Files.exists(arthasDir)) {
+            Files.createDirectory(arthasDir);
+        }
+
+        Runnable r;
+        if (!Files.exists(toArthasJar(arthasDir))) {
+            r = () -> startArthas(downloadArthas(arthasDir, arthasVersion), arthasHttpPort);
+        } else {
+            r = () -> startArthas(toArthasJar(arthasDir), arthasHttpPort);
+        }
+        var arthasThread = new Thread(r);
+        arthasThread.setName("Arthas thread");
+        arthasThread.setDaemon(true);
+        arthasThread.start();
+    }
+
+    private Path downloadArthas(Path arthasDir, String version) {
+        var downloadURL = "https://arthas.aliyun.com/download/" + version;
+        try {
+            log.debug("Attempting to download Arthas from " + downloadURL);
+            var arthasZipFile = File.createTempFile("arthas-" + version, ".zip");
+            try (BufferedInputStream in = new BufferedInputStream(new URL(downloadURL).openStream());
+                    FileOutputStream fileOutputStream = new FileOutputStream(arthasZipFile)) {
+                byte[] dataBuffer = new byte[10240];
+                int bytesRead;
+                while ((bytesRead = in.read(dataBuffer, 0, 10240)) != -1) {
+                    fileOutputStream.write(dataBuffer, 0, bytesRead);
+                }
+            }
+            ZipUtils.unzip(arthasZipFile.toPath(), arthasDir);
+            log.debug("Arthas downloaded");
+            return toArthasJar(arthasDir);
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private void startArthas(Path arthasJar, Integer httpPort) {
+        log.debug("Attempting to start Arthas");
+        ProcessBuilder processBuilder = new ProcessBuilder("java",
```

Comment:
I remember some more complicated logic to get the right java binary somewhere in the code base but maybe it's just a bad recollection.

Reply:
Right, we do have that. Let me see if I can find it

Reply:
Is it in `io.quarkus.utilities.JavaBinFinder`?

Reply:
Fixed